File: vulnerable_app.py
import os
import subprocess
import sqlite3
import pickle
import hashlib
import yaml
import xml.etree.ElementTree as ET
import random
import tempfile

# Hardcoded credentials (CWE-798)
DB_PASSWORD = "supersecret123"
AWS_ACCESS_KEY = "AKIAIOSFODNN7EXAMPLE"
AWS_SECRET_KEY = "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
SECRET_TOKEN = "ghp_hardcodedGitHubToken1234567890abcdef"

# SQL Injection (CWE-89)
def get_user(username):
    conn = sqlite3.connect("users.db")
    cursor = conn.cursor()
    query = "SELECT * FROM users WHERE username = '" + username + "'"
    cursor.execute(query)
    return cursor.fetchall()

def get_order(order_id):
    conn = sqlite3.connect("orders.db")
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM orders WHERE id = %s" % order_id)
    return cursor.fetchall()

# Command Injection (CWE-78)
def ping_host(host):
    os.system("ping -c 1 " + host)

def run_command(cmd):
    subprocess.call(cmd, shell=True)

def list_files(directory):
    return os.popen("ls " + directory).read()

# Path Traversal (CWE-22)
def read_file(filename):
    with open("/var/app/" + filename, "r") as f:
        return f.read()

def download_file(path):
    base = "/uploads/"
    return open(base + path).read()

# Insecure Deserialization (CWE-502)
def load_data(data):
    return pickle.loads(data)

def load_user_object(serialized):
    return pickle.loads(serialized)

# XSS - reflected (CWE-79)
def render_page(user_input):
    return "<html><body>" + user_input + "</body></html>"

def greet_user(name):
    return f"<h1>Hello {name}</h1>"

# Weak cryptography (CWE-326, CWE-327)
def hash_password(password):
    return hashlib.md5(password.encode()).hexdigest()

def encrypt_data(data):
    return hashlib.sha1(data.encode()).hexdigest()

# XML External Entity (XXE) (CWE-611)
def parse_xml(xml_string):
    return ET.fromstring(xml_string)

def load_config(xml_file):
    tree = ET.parse(xml_file)
    return tree.getroot()

# Insecure random for security use (CWE-330)
def generate_session_token():
    return str(random.randint(0, 999999))

def generate_reset_token():
    return hex(random.getrandbits(64))

# YAML deserialization (CWE-502)
def load_yaml_config(data):
    return yaml.load(data)  # unsafe load

# Insecure temp file (CWE-377)
def create_temp():
    return tempfile.mktemp()

# Open redirect (CWE-601)
def redirect_user(url):
    return f"Location: {url}"

# Sensitive data in logs (CWE-532)
def login(user, password):
    print(f"Login attempt: user={user} password={password}")
    return password == DB_PASSWORD

# SSRF (CWE-918)
def fetch_url(url):
    import urllib.request
    return urllib.request.urlopen(url).read()

# Eval injection (CWE-95)
def calculate(expression):
    return eval(expression)

def run_code(code):
    exec(code)

# Integer overflow-like logic
def allocate_buffer(size):
    return bytearray(size * 1024 * 1024 * 1024)

if __name__ == "__main__":
    get_user("admin' OR '1'='1")
    ping_host("google.com; rm -rf /")
    read_file("../../etc/passwd")
    print(hash_password("password"))
    print(generate_session_token())
    calculate("__import__('os').system('id')")

File: vulnerable_app.rb
require 'open3'
require 'digest'
require 'yaml'
require 'net/http'
require 'sqlite3'
require 'json'

# Hardcoded credentials (CWE-798)
DB_PASSWORD = "ruby_hardcoded_pass_123"
AWS_KEY = "AKIAIOSFODNN7RBEXAMPLE"
STRIPE_SECRET = "sk_live_hardcoded_stripe_key_1234567890"

# SQL Injection (CWE-89)
def get_user(username)
  db = SQLite3::Database.new("users.db")
  db.execute("SELECT * FROM users WHERE username = '#{username}'")
end

def get_order(id)
  db = SQLite3::Database.new("orders.db")
  db.execute("SELECT * FROM orders WHERE id = " + id)
end

def delete_record(table, id)
  db = SQLite3::Database.new("app.db")
  db.execute("DELETE FROM #{table} WHERE id = #{id}")
end

# Command Injection (CWE-78)
def ping_host(host)
  system("ping -c 1 #{host}")
end

def run_command(cmd)
  `#{cmd}`
end

def list_files(dir)
  Open3.capture2("ls #{dir}")
end

# Path Traversal (CWE-22)
def read_file(filename)
  File.read("/var/app/#{filename}")
end

def write_file(path, content)
  File.write("/uploads/#{path}", content)
end

# Weak cryptography (CWE-326)
def hash_password(password)
  Digest::MD5.hexdigest(password)
end

def weak_hash(data)
  Digest::SHA1.hexdigest(data)
end

# Insecure deserialization (CWE-502)
def load_object(data)
  Marshal.load(data)
end

def load_yaml_config(data)
  YAML.load(data)  # unsafe, allows object instantiation
end

# Insecure random (CWE-330)
def generate_token
  rand(999999).to_s
end

def generate_otp
  rand(9000) + 1000
end

# SSRF (CWE-918)
def fetch_url(url)
  Net::HTTP.get(URI(url))
end

# XSS (CWE-79)
def greet_user(name)
  "<h1>Hello #{name}</h1>"
end

# Open redirect (CWE-601)
def redirect_to(url)
  "Location: #{url}"
end

# Code injection via eval (CWE-95)
def calculate(expression)
  eval(expression)
end

def run_code(code)
  eval(code)
end

# Sensitive data in logs (CWE-532)
def login(user, password)
  puts "Login attempt: user=#{user} password=#{password}"
  password == DB_PASSWORD
end

# RegEx DoS (CWE-1333)
def validate_email(email)
  email.match(/^([a-zA-Z0-9])+([a-zA-Z0-9\._-])*@([a-zA-Z0-9_-])+([a-zA-Z0-9\._-]+)+$/)
end

# Mass assignment vulnerability
def update_user(params)
  User.update(params)  # All params passed directly
end

get_user("admin' OR '1'='1'--")
ping_host("google.com; cat /etc/passwd")
read_file("../../etc/shadow")
puts hash_password("password")
puts calculate("system('id')")
